schnorr_fun/src/frost_backup.rs

```diff
@@ -0,0 +1,198 @@
+//!
+//! 32^4 = 2^20 ~> a collision rate of about one in a million.
+//!
+//! ### Secret Share - (52)
+//! A FROST secret share is fixed length scalar of 32 bytes.
+//! This is 32 * 8 / 5 = 51.2 -> 52 bech32 characters
+//!
+//! ### Share Index - (1 or 52)
+//! The share index can be be scalar, but will often be small for simplicity and smaller backups.
+//! By leaving this data piece at the end, we can use the length of the remaining data to
+//! easily decode a single bech32 char into integer, or 32 chars into a scalar.
+use alloc::{string::String, vec::Vec};
+use bech32::{self, u5, FromBase32, ToBase32, Variant::Bech32m};
+use secp256kfun::{
+    digest::{generic_array::typenum::U32, Digest},
+    marker::Secret,
+    Point, Scalar,
+};
+
+/// Secret shares can have a small index up to 32, or index at a specific scalar.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum BackupShareIndex {
+    /// a secret share at a small index up to 32
+    SmallIndex(usize),
+    /// a secret share at a specific scalar index
+    Scalar(Scalar),
+}
+
+/// An error encountered when encoding a Frostsnap backup.
+#[derive(Debug, Copy, Clone)]
+pub enum FrostBackupError {
+    /// Generic error from bech32 library
+    Bech32Error(bech32::Error),
+    /// Threshold was zero
+    ThresholdCantBeZero,
+    /// Threshold was greater than 32.
+    ThresholdTooBig,
+    /// Tried to decode a bech32 variant that was not bech32m
+    WrongBech32Variant(bech32::Variant),
+    /// Decoded secret share is not a valid secp256k1 scalar
+    InvalidSecretShareScalar,
+    /// Tried to decode backup with unknown prefix
+    InvalidHumanReadablePrefix,
+    /// The share index data length doesn't match expected for SmallIndex nor Scalar
+    UnknownShareIndexType,
+    /// Decoded share index is not a valid secp256k1 scalar
+    InvalidShareIndexScalar,
+}
+
+/// Create an identifier that's used to determine compatibility of FROST secret shares.
+/// The first 4 bech32 chars from a hash of the polynomial coefficients.
+/// Collision expected once in (32)^4 = 2^20.
+pub fn polynomial_identifier<H: Default + Digest<OutputSize = U32>>(
+    polynomial: Vec<Point>,
+) -> [u5; 4] {
+    let mut hash = H::default();
+    for coefficient in polynomial.into_iter() {
+        hash.update(coefficient.to_bytes());
+    }
+    hash.finalize().to_vec().to_base32()[0..4]
+        .try_into()
+        .expect("4 bech32 chars must fit 4 character arry")
+}
+
+/// Create a bech32m FROST backup
+///
+/// Requires that the threshold be no greater than 32.
+/// If using an integer (small) index, the backup will be 67 bech32 characters.
+/// If using a scalar index, the backup will be 118 bech32 characters.
+pub fn encode_backup<H: Default + Digest<OutputSize = U32>>(
+    threshold: usize,
+    polynomial: Vec<Point>,
+    secret_share: Scalar<Secret>,
+    share_index: BackupShareIndex,
+) -> Result<String, FrostBackupError> {
```

Comment:
Encoding and decoding errors should be different. I don't know if there is any justifiable reason to have an error for encoding. Maybe we want to make sure the threshold fits but everything else seems like a programmer error (can just be a panic). Rather than return an error you could have a `ShortThreshold` type you force the caller to construct it to make them better aware of the constraint. Or `NonZeroU5` since you can also use it for the party index.